### PR TITLE
feat: bump driver to 4.x beta COMPASS-4805

### PR DESCRIPTION
**Pull request by addaleax**
This is fairly functional already, but ~~e.g. the aggregation pipeline builder doesn't appear to work yet, and~~ I have only checked a small subset of Compass's functionality here.

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [x] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [x] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by mcasimir** [comment]
> the aggregation pipeline builder doesn't appear to work yet

What does it mean? :)

File: packages/connection-model/package.json [hunk 1]
```diff
@@ -23,14 +23,17 @@
...
+  },
   "dependencies": {
     "@mongodb-js/ssh-tunnel": "^1.2.0",
     "ampersand-model": "^8.0.0",
     "ampersand-rest-collection": "^6.0.0",
     "async": "^3.1.0",
     "debug": "^4.1.1",
     "lodash": "^4.17.15",
-    "mongodb": "^3.6.3",
+    "mongodb3": "npm:mongodb@^3.6.3",
```
Comment (mcasimir) [inline packages/connection-model/package.json:36]: Could we 'copy-paste' the `parse-connection-string.js` file and drop this one? We need to rework that to switch it to the new module anyway.

**Comment by addaleax** [comment]
> > the aggregation pipeline builder doesn't appear to work yet
> 
> What does it mean? :)

@mcasimir It means that the error from the driver could be `undefined` instead of `null` for aggregation .toArray(), but that's fixed now (in bbd05d0dbbbe5694838662bdf028f001c9b95080) :)

File: packages/data-service/lib/native-client.js [hunk 2]
```diff
@@ -131,14 +131,14 @@ class NativeClient extends EventEmitter {
...
-    this._isConnected = true;
+    this._isConnecting = true;
 
     connect(
       this.model,
       this.setupListeners.bind(this),
       (err, _client, tunnel, connectionOptions) => {
         if (err) {
-          this._isConnected = false;
+          this._isConnecting = false;
```
Comment (mcasimir) [inline packages/data-service/lib/native-client.js:141]: Nice :)

**Comment by mcasimir** [comment]
> > > the aggregation pipeline builder doesn't appear to work yet
> > 
> > 
> > What does it mean? :)
> 
> @mcasimir It means that the error from the driver could be `undefined` instead of `null` for aggregation .toArray(), but that's fixed now (in [bbd05d0](https://github.com/mongodb-js/compass/commit/bbd05d0dbbbe5694838662bdf028f001c9b95080)) :)

Oh ok :)

**Review by mcasimir — APPROVED** [review]
Great! Should we merge now and get this in the next release? Or we prefer to wait?

File: packages/connection-model/package.json (hunk 1 above)
Comment (addaleax) [reply packages/connection-model/package.json:36]: I think we could... but I don't think we'd exactly want that to be the long-term solution here either, it's probably best to do some actual URL parsing based on mongodb-connection-string-url and then actually handle all the transformations here in the connection model directly (assuming that we keep mongodb-connection-model)

**Comment by addaleax** [comment]
> Great! Should we merge now and get this in the next release? Or we prefer to wait?

As you prefer :) In any case, we should do at least one beta with this I would say. And I don't know if we want to actually do a compass GA release with a beta of the driver instead of a rc.0?

File: packages/connection-model/package.json (hunk 1 above)
Comment (mcasimir) [reply packages/connection-model/package.json:36]: > but I don't think we'd exactly want that to be the long-term solution here either

Yeah i was thinking the same, we should definitely use the `mongodb-connection-string-url` module instead.

**Comment by mcasimir** [comment]
> > Great! Should we merge now and get this in the next release? Or we prefer to wait?
> 
> As you prefer :) In any case, we should do at least one beta with this I would say. And I don't know if we want to actually do a compass GA release with a beta of the driver instead of a rc.0?

There should be no issue with that. What does @mmarcon says?

**Comment by mmarcon** [comment]
We are going to GA the shell with a beta of the driver, I think it's fine to ship Compass too.

But yeah, let's definitely do a beta of compass with that before we do a GA.

**Pull request merged**